File: platforms/android/feedback/src/main/kotlin/io/markerusa/feedback/Consent.kt
package io.markerusa.feedback

/**
 * Consent, mirroring `packages/core/src/consent.ts`.
 *
 * Same rule as web: capture is gated on an explicit, current grant, and every
 * uncertain path resolves to "not granted". A missed capture is a product
 * problem; an unconsented one is a legal problem.
 */

enum class ConsentScope(val wire: String) {
    SCREENSHOT("screenshot"),
    DIAGNOSTICS("diagnostics"),
    SESSION_REPLAY("session_replay")
}

data class ConsentRecord(
    val scopes: List<ConsentScope>,
    val policyVersion: String,
    val grantedAt: Long,
    val source: String
)

class ConsentManager(
    private val storage: KeyValueStore,
    private val policyVersion: String
) {
    private companion object {
        const val KEY_SCOPES = "markerio.consent.scopes"
        const val KEY_VERSION = "markerio.consent.version"
        const val KEY_GRANTED_AT = "markerio.consent.grantedAt"
        const val KEY_SOURCE = "markerio.consent.source"
    }

    /**
     * Stored as discrete keys rather than a serialized blob specifically because
     * this class has no JSON parser. Keeping consent readable without one means
     * the gate can never fail open due to a deserialization bug.
     */
    fun load(): ConsentRecord? {
        val version = storage.get(KEY_VERSION) ?: return null
        // A grant against superseded terms is not a grant against these ones.
        if (version != policyVersion) return null

        val raw = storage.get(KEY_SCOPES) ?: return null
        val scopes = raw.split(',')
            .filter { it.isNotBlank() }
            .mapNotNull { wire -> ConsentScope.entries.firstOrNull { it.wire == wire } }
        if (scopes.isEmpty()) return null

        return ConsentRecord(
            scopes = scopes,
            policyVersion = version,
            grantedAt = storage.get(KEY_GRANTED_AT)?.toLongOrNull() ?: 0L,
            source = storage.get(KEY_SOURCE) ?: "host_app"
        )
    }

    /** The gate every capture path calls. */
    fun has(scope: ConsentScope): Boolean = load()?.scopes?.contains(scope) == true

    fun grant(scopes: List<ConsentScope>, source: String): ConsentRecord {
        // Union with any prior grant, so granting replay cannot silently drop
        // screenshot/diagnostics.
        val merged = ((load()?.scopes ?: emptyList()) + scopes).distinct()
        val now = System.currentTimeMillis()

        storage.set(KEY_SCOPES, merged.joinToString(",") { it.wire })
        storage.set(KEY_VERSION, policyVersion)
        storage.set(KEY_GRANTED_AT, now.toString())
        storage.set(KEY_SOURCE, source)

        return ConsentRecord(merged, policyVersion, now, source)
    }

    /** Withdrawal must be as easy as granting, and takes effect immediately. */
    fun revoke() {
        storage.remove(KEY_SCOPES)
        storage.remove(KEY_VERSION)
        storage.remove(KEY_GRANTED_AT)
        storage.remove(KEY_SOURCE)
    }
}

File: platforms/ios/Sources/SusaSignals/Consent.swift
import Foundation

/**
 Consent, mirroring `packages/core/src/consent.ts` and the Android implementation.

 Same rule on every platform: capture is gated on an explicit, current grant, and
 every uncertain path resolves to "not granted". A missed capture is a product
 problem; an unconsented one is a legal problem.
 */

public enum ConsentScope: String, Codable, CaseIterable {
    case screenshot
    case diagnostics
    case sessionReplay = "session_replay"
    /// A separate legal basis from diagnostics under GDPR and ePrivacy, so it
    /// cannot ride on the implicit consent that filing a report carries.
    case analytics
}

public struct ConsentRecord: Codable, Equatable {
    public let scopes: [ConsentScope]
    public let policyVersion: String
    public let grantedAt: Int64
    public let source: String
}

public final class ConsentManager {
    private enum Key {
        static let scopes = "susa.signals.consent.scopes"
        static let version = "susa.signals.consent.version"
        static let grantedAt = "susa.signals.consent.grantedAt"
        static let source = "susa.signals.consent.source"
    }

    private let storage: KeyValueStore
    private let policyVersion: String

    public init(storage: KeyValueStore, policyVersion: String) {
        self.storage = storage
        self.policyVersion = policyVersion
    }

    /// Stored as discrete keys rather than a serialized blob so the gate can never
    /// fail open because of a deserialization bug.
    public func load() -> ConsentRecord? {
        guard let version = storage.get(Key.version) else { return nil }
        // A grant against superseded terms is not a grant against these ones.
        guard version == policyVersion else { return nil }

        guard let raw = storage.get(Key.scopes) else { return nil }
        let scopes = raw
            .split(separator: ",")
            .map(String.init)
            .compactMap(ConsentScope.init(rawValue:))
        guard !scopes.isEmpty else { return nil }

        return ConsentRecord(
            scopes: scopes,
            policyVersion: version,
            grantedAt: Int64(storage.get(Key.grantedAt) ?? "") ?? 0,
            source: storage.get(Key.source) ?? "host_app"
        )
    }

    /// The gate every capture path calls.
    public func has(_ scope: ConsentScope) -> Bool {
        load()?.scopes.contains(scope) ?? false
    }

    @discardableResult
    public func grant(_ scopes: [ConsentScope], source: String) -> ConsentRecord {
        // Union with any prior grant, so granting replay cannot silently drop
        // screenshot/diagnostics.
        var merged = load()?.scopes ?? []
        for scope in scopes where !merged.contains(scope) {
            merged.append(scope)
        }

        let now = Int64(Date().timeIntervalSince1970 * 1000)
        storage.set(Key.scopes, merged.map(\.rawValue).joined(separator: ","))
        storage.set(Key.version, policyVersion)
        storage.set(Key.grantedAt, String(now))
        storage.set(Key.source, source)

        return ConsentRecord(scopes: merged, policyVersion: policyVersion, grantedAt: now, source: source)
    }

    /// Withdrawal must be as easy as granting, and takes effect immediately.
    public func revoke() {
        storage.remove(Key.scopes)
        storage.remove(Key.version)
        storage.remove(Key.grantedAt)
        storage.remove(Key.source)
    }
}
